src/compiler/checker.ts

```diff
@@ -29471,14 +29471,25 @@ namespace ts {
                 for (let i = spreadIndex; i < args.length; i++) {
                     const arg = args[i];
                     // We can call checkExpressionCached because spread expressions never have a contextual type.
-                    const spreadType = arg.kind === SyntaxKind.SpreadElement && (flowLoopCount ? checkExpression((arg as SpreadElement).expression) : checkExpressionCached((arg as SpreadElement).expression));
-                    if (spreadType && isTupleType(spreadType)) {
-                        forEach(getTypeArguments(spreadType), (t, i) => {
-                            const flags = spreadType.target.elementFlags[i];
-                            const syntheticArg = createSyntheticExpression(arg, flags & ElementFlags.Rest ? createArrayType(t) : t,
-                                !!(flags & ElementFlags.Variable), spreadType.target.labeledElementDeclarations?.[i]);
-                            effectiveArgs.push(syntheticArg);
-                        });
+                    const argType = arg.kind === SyntaxKind.SpreadElement && (flowLoopCount ? checkExpression((arg as SpreadElement).expression) : checkExpressionCached((arg as SpreadElement).expression));
+                    if (argType) {
+                        const spreadTypes = !!(argType.flags & TypeFlags.Intersection) ? (argType as IntersectionType).types : [argType];

```

Comment:
just a guess, but I think that the code might be easier to read if you handled intersections separately from the other case. Something like

```ts
let spreadType
if (intersection)
  find the spread type (or undefined if not exactly one)
else
  spreadType = isTupleType(argType) ? argType : undefined
if (spreadType) {
  forEach(getTypeArguments(spreadType), ...
```

Reply:
I've switched to a single loop as suggested.